### PR TITLE
gh-119180: Add `annotationlib` module to support PEP 649

### PR DESCRIPTION
**Pull request by JelleZijlstra**
This PR implements the `annotationlib` module proposed by PEP-749, as well as related Python changes for PEP-649 and PEP-749.

* Issue: gh-119180

### CODE REVIEW

File: Doc/howto/descriptor.rst [hunk 1]
```diff
@@ -1366,6 +1366,10 @@ Using the non-data descriptor protocol, a pure Python version of
         def __call__(self, *args, **kwds):
             return self.f(*args, **kwds)
 
+        @property
+        def __annotations__(self):
+            return self.f.__annotations__
```
Comment (AlexWaygood) [inline Doc/howto/descriptor.rst:1371]: You need to delete the reference to `~function.__annotations__` in the following paragraph if you want to make this change

File: Lib/annotationlib.py [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+        )
+
+    @property
+    def __forward_arg__(self):
+        if self.__arg__ is not None:
+            return self.__arg__
+        if self.__ast_node__ is not None:
+            self.__arg__ = ast.unparse(self.__ast_node__)
+            return self.__arg__
+        raise RuntimeError("Forward reference is not initialized")
```
Comment (AlexWaygood) [inline Lib/annotationlib.py]: I have a pretty strong dislike of properties that raise non-`AttributeError` exceptions, because they break the expectations of common idioms like `forward_arg = getattr(obj, "__forward_arg__", None)` :/

The caller generally expects that code to be safe, not anticipating that the `getattr()` call will only suppress `AttributeError`s.

File: Lib/typing.py [hunk 3]
```diff
@@ -2187,7 +2168,7 @@ class _AnnotatedAlias(_NotIterable, _GenericAlias, _root=True):
     """Runtime representation of an annotated type.
 
     At its core 'Annotated[t, dec1, dec2, ...]' is an alias for the type 't'
-    with extra annotations. The alias behaves like a normal typing alias.
+    with extra annotationlib. The alias behaves like a normal typing alias.
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
    with extra metadata. The alias behaves like a normal typing alias.
```

**Review by AlexWaygood — COMMENTED** [review]
Nice. Haven't looked in-depth yet, but a few things I noticed off the bat from skimming:

File: Lib/annotationlib.py (hunk 2 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py]: This particular condition should never happen, because a properly initialized ForwardRef will always have one of the two attributes, so I do think RuntimeError (or AssertionError) is appropriate.

Comment (AlexWaygood) [reply Lib/annotationlib.py]: Okay, in that case I think this would communicate that more clearly:

```suggestion
        raise AssertionError(
            "Attempted to access '__forward_arg__' on an uninitialized ForwardRef"
        )
```

File: Lib/annotationlib.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
+"""Helpers for introspecting and wrapping annotations."""
+
+import ast
+import enum
+import functools
+import sys
+import types
+
+
```
Comment (picnixz) [inline Lib/annotationlib.py:11]: Can you perhaps add an `__all__` export? Because otherwise, I would think that `ForwardRef` is public but your comment tells me it's an "internal" wrapper, so it could be confusing.

File: Lib/annotationlib.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+    SOURCE = 3
+
+
+_Union = None
+_sentinel = object()
+
+# Slots shared by ForwardRef and _Stringifier. The __forward__ names must be
+# preserved for compatibility with the old typing.ForwardRef class. The remaining
+# names are private.
+_slots = (
```
Comment (picnixz) [inline Lib/annotationlib.py]:
```suggestion
_PRIVATE_SLOTS = (
```

maybe it's easier to spot that it's a constant that way

File: Lib/annotationlib.py [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+            )
+        else:
+            return ast.Constant(value=other)
+
+    def __make_new(self, node):
+        return _Stringifier(
+            node, self.__globals__, self.__owner__, self.__forward_is_class__
+        )
+
+    def _make_binop(op: ast.AST):
```
Comment (picnixz) [inline Lib/annotationlib.py:314]: I would maybe add a comment just above to indicate that those are no more methods to the class, but temporary helpers that will be deleted just after usage.

File: Lib/annotationlib.py [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+    __ge__ = _make_compare(ast.GtE())
+
+    del _make_compare
+
+    # Doesn't work because the return type is always coerced to a bool
+    # __contains__ = _make_compare(ast.In())
+
+    # Must implement this since we set __eq__. We hash by identity so that
+    # stringifiers in dict keys are kept separate.
+    def __hash__(self):
```
Comment (picnixz) [inline Lib/annotationlib.py]: I would put that one before the binary oeprators.

File: Lib/annotationlib.py [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+
+        return unary_op
+
+    __invert__ = _make_unary_op(ast.Invert())
+    __pos__ = _make_unary_op(ast.UAdd())
+    __neg__ = _make_unary_op(ast.USub())
+
+    del _make_unary_op
+
+    def __getitem__(self, other):
```
Comment (picnixz) [inline Lib/annotationlib.py]: Explicit methods should perhaps come first, then followed by all those created by `_make*` calls?

File: Lib/annotationlib.py [hunk 9]
```diff
@@ -0,0 +1,635 @@
+"""Helpers for introspecting and wrapping annotations."""
+
+import ast
+import enum
+import functools
+import sys
+import types
+
+
+class Format(enum.IntEnum):
```
Comment (picnixz) [inline Lib/annotationlib.py:12]: So you eventually decided not to have a global enumeration?

File: Lib/annotationlib.py [hunk 10]
```diff
@@ -0,0 +1,635 @@
...
+            except ValueError:
+                pass
+            else:
+                self.__forward_evaluated__ = True
+                self.__forward_value__ = value
+                return value
+        if owner is None:
+            owner = self.__owner__
+        if owner is None and type_params is None:
+            raise TypeError("Either 'owner' or 'type_params' must be provided")
```
Comment (picnixz) [inline Lib/annotationlib.py]: This may seem un-necessary, but I would have expected the 'type_params' or 'owner' instead to match the signature.

File: Lib/annotationlib.py [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+                globals = getattr(owner, "__globals__", None)
+        if self.__forward_module__ is not None:
+            globals = getattr(
+                sys.modules.get(self.__forward_module__, None), "__dict__", globals
+            )
+
+        if locals is None:
+            locals = {}
+            if isinstance(self.__owner__, type):
+                locals.update(vars(self.__owner__))
```
Comment (picnixz) [inline Lib/annotationlib.py:116]: This is something that I never thought about, but *every* type supports `__dict__` or is there some really weird constructions that could not have one?

File: Lib/typing.py [hunk 12]
```diff
@@ -1011,102 +1018,76 @@ def run(arg: Child | Unrelated):
...
+    locals=None,
+    type_params=None,
+    format=annotationlib.Format.VALUE,
+    _recursive_guard=frozenset(),
+):
+    """Evaluate a forward reference as a type hint.
+
+    This is similar to calling the ForwardRef.evaluate() method,
+    but unlike that method, evaluate_forward_ref() also:
+    * Recursively evaluates forward references nested within the type hint.
```
Comment (picnixz) [inline Lib/typing.py]:
```suggestion
    but unlike that method, evaluate_forward_ref() also:
    
    * Recursively evaluates forward references nested within the type hint.
```

File: Misc/NEWS.d/next/Library/2024-06-11-07-17-25.gh-issue-119180.iH-2zy.rst [hunk 13]
```diff
@@ -0,0 +1,4 @@
+As part of implementing :pep:`649` and :pep:`749`, add a new module
+``annotationlib``. Add support for unresolved forward references in
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-06-11-07-17-25.gh-issue-119180.iH-2zy.rst:2]:
```suggestion
:mod:`annotationlib`. Add support for unresolved forward references in
```

File: Lib/functools.py [hunk 14]
```diff
@@ -889,13 +889,19 @@ def register(cls, func=None):
 
             # only import typing if annotation parsing is necessary
             from typing import get_type_hints
-            argname, cls = next(iter(get_type_hints(func).items()))
+            from annotationlib import Format, ForwardRef
+            argname, cls = next(iter(get_type_hints(func, format=Format.FORWARDREF).items()))
```
Comment (AlexWaygood) [inline Lib/functools.py:897]: We could possibly fix #110373 at the same time here by using `typing.evaluate_forward_ref()` instead of `get_type_hints`. `singledispatch` only needs to evaluate the first type hint on a function; anything else is doing extra work and leads to bugs.

(Ideally we'd just use `annotationlib.ForwardRef.evaluate()` and then we wouldn't have to import `typing` at all here, but that feels risky and possibly not backwards compatible, given all the transformations that `get_type_hints` does.)

File: Lib/annotationlib.py (hunk 4 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py:11]: I will add `__all__`. `ForwardRef` is public, I'll adjust the docstring.

File: Lib/annotationlib.py (hunk 5 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py]: Will make it uppercase, I don't think we need the `_PRIVATE` part.

File: Lib/annotationlib.py (hunk 7 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py]: Rearranging them

File: Lib/annotationlib.py (hunk 9 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py:12]: It's in PEP 749 this way.

File: Lib/annotationlib.py (hunk 11 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py:116]: It is possible in a subclass of `type` I believe.

File: Misc/NEWS.d/next/Library/2024-06-11-07-17-25.gh-issue-119180.iH-2zy.rst (hunk 13 above)
Comment (JelleZijlstra) [reply Misc/NEWS.d/next/Library/2024-06-11-07-17-25.gh-issue-119180.iH-2zy.rst:2]: This doesn't work yet because this PR doesn't add documentation for the new module. I'll do that in a separate module.

File: Lib/annotationlib.py (hunk 6 above)
Comment (JelleZijlstra) [reply Lib/annotationlib.py:314]: I think that's obvious from the code.

File: Lib/annotationlib.py (hunk 11 above)
Comment (picnixz) [reply Lib/annotationlib.py:116]: Should you guard it against such cases? or do you just let the exception propagate?

File: Lib/test/test_annotationlib.py [hunk 15]
```diff
@@ -0,0 +1,647 @@
...
+        self.assertEqual(annotationlib.get_annotations(C), {"x": int})
+
+    def test_missing_dunder_dict(self):
+        class NoDict(type):
+            @property
+            def __dict__(cls):
+                raise AttributeError
+
+        class C1(metaclass=NoDict):
+            a: int
```
Comment (picnixz) [inline Lib/test/test_annotationlib.py]: This one still has a `__dict__`, so you should instead use something like:

```python
class NoDict(type):
    def __getattr__(self, item):
        raise AttributeError
    
    def __getattribute__(self, item):
        if item == '__dict__':
            raise AttributeError
        return super().__getattribute__(item)

class A(metaclass=NoDict):
	pass
```

Then, `vars(A)` would raises an error (the first '.__dict__' gets into `__getattribute__`, and *then* it gets into `__getattr__` actually so you want to suppress them both).

(Btw, I'm sorry to bother you with that, but with all the weird stuff that I've seen in Sphinx and the interfaces that have weird constructions, when it comes to typing I'd prefer having something that is robust and covers some weird cases).

Comment (JelleZijlstra) [reply Lib/test/test_annotationlib.py]: `vars(C1)` raises an error here already:

```
>>> class NoDict(type):
...     @property
...     def __dict__(cls):
...         raise AttributeError
... 
... class C1(metaclass=NoDict):
...     a: int
... 
>>> vars(C1)
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    vars(C1)
    ~~~~^^^^
TypeError: vars() argument must have __dict__ attribute
```

However, the tests I added don't yet cover cases where we call `vars()` on the "owner", I'll work on that later.

Comment (picnixz) [reply Lib/test/test_annotationlib.py]: Oh! this behaviour changed between two versions actually! (I incorrectly checked it with 3.12.2 actually and not 3.14!)

Comment (AlexWaygood) [reply Lib/test/test_annotationlib.py]: > Oh! this behaviour changed between two versions actually! (I incorrectly checked it with 3.12.2 actually and not 3.14!)

Um, that sounds bad. We should try to bisect that and make sure it was deliberate.

Comment (picnixz) [reply Lib/test/test_annotationlib.py]: Oh no false alarm. How come I got no error yesterday but I got one today? maybe it's an even older version. Let me verify again.

EDIT: Sorry Alex, what I said on the behaviour was incorrect!

Comment (picnixz) [reply Lib/test/test_annotationlib.py]: Ah ! I know, actually I was trying to make the error on the *type* of `C1`, namely `vars(type(C1))` and not on `C1` itself. However, it should be possible to have *owner = NoDict* or is it not possible?

Comment (JelleZijlstra) [reply Lib/test/test_annotationlib.py]: This actually made me find an even more basic bug:

```
>>> class X(type):
...     a: int
...     
>>> class Y(metaclass=X):
...     b: str
...     
>>> Y.__annotations__
{'a': <class 'int'>}
```

Comment (JelleZijlstra) [reply Lib/test/test_annotationlib.py]: Posted https://discuss.python.org/t/pep-749-implementing-pep-649/54974/28 about this, I think I'll want to deal with it in a separate PR (since the solution likely involves changes to #119361).

File: Lib/functools.py (hunk 14 above)
Comment (JelleZijlstra) [reply Lib/functools.py:897]: I think I'll leave that for another PR, added it to #119180.

File: Lib/test/test_annotationlib.py (hunk 15 above)
Comment (JelleZijlstra) [reply Lib/test/test_annotationlib.py]: For this PR I won't handle the metaclass case yet; still thinking about how to handle that.

**Comment by JelleZijlstra** [comment]
This PR is still ready for review. While we will likely make some change related to metaclasses, that won't materially effect the changes here. I'd like to get this PR landed as a foundation to build the rest of the PEP implementation on, so I'd appreciate any reviews.

File: Lib/annotationlib.py [hunk 16]
```diff
@@ -0,0 +1,655 @@
...
+                    if module:
+                        globals = getattr(module, "__dict__", None)
+            elif isinstance(owner, types.ModuleType):
+                globals = getattr(owner, "__dict__", None)
+            elif callable(owner):
+                globals = getattr(owner, "__globals__", None)
+        if self.__forward_module__ is not None:
+            globals = getattr(
+                sys.modules.get(self.__forward_module__, None), "__dict__", globals
+            )
```
Comment (carljm) [inline Lib/annotationlib.py]: Since this (if set) will override the above heuristics on lines 97-107 anyway, why not move it above line 97 so that if we get a globals from `__forward_module__` we don't need to bother with those other lookups?

File: Lib/test/test_annotationlib.py [hunk 17]
```diff
@@ -0,0 +1,771 @@
...
+        def f(x: int, y: doesntexist):
+            pass
+
+        anno = annotationlib.get_annotations(f, format=annotationlib.Format.SOURCE)
+        self.assertEqual(anno, {"x": "int", "y": "doesntexist"})
+
+    def test_expressions(self):
+        def f(
+            add: a + b,
+            sub: a + b,
```
Comment (carljm) [inline Lib/test/test_annotationlib.py]:
```suggestion
            sub: a - b,
```

File: Lib/test/test_annotationlib.py [hunk 18]
```diff
@@ -0,0 +1,771 @@
...
+            *args: *a,
+        ):
+            pass
+
+        anno = annotationlib.get_annotations(f, format=annotationlib.Format.SOURCE)
+        self.assertEqual(
+            anno,
+            {
+                "add": "a + b",
+                "sub": "a + b",
```
Comment (carljm) [inline Lib/test/test_annotationlib.py]:
```suggestion
                "sub": "a - b",
```

File: Lib/test/test_annotationlib.py [hunk 19]
```diff
@@ -0,0 +1,771 @@
...
+            with self.assertRaises(TypeError):
+                pickle.dumps(fr, proto)
+
+
+class TestGetAnnotations(unittest.TestCase):
+    def test_builtin_type(self):
+        self.assertEqual(annotationlib.get_annotations(int), {})
+        self.assertEqual(annotationlib.get_annotations(object), {})
+
+    def test_custom_metaclass(self):
```
Comment (carljm) [inline Lib/test/test_annotationlib.py:257]: Do you plan to add tests here for the pathological cases with metaclass annotations, or make that change in a separate PR?

File: Lib/test/test_annotationlib.py [hunk 20]
```diff
@@ -0,0 +1,771 @@
...
+
+        with self.assertRaises(ValueError):
+            annotationlib.get_annotations(f1, format=0)
+
+        with self.assertRaises(ValueError):
+            annotationlib.get_annotations(f1, format=4)
+
+    def test_custom_object_with_annotations(self):
+        class C:
+            def __init__(self, x: int = 0, y: str = ""):
```
Comment (carljm) [inline Lib/test/test_annotationlib.py]: Are the arguments and their annotations actually related to the test? If not, I'd remove them for clarity about what behavior the test is actually specifying.

**Review by carljm — APPROVED** [review]
This looks good to me!

File: Lib/test/test_annotationlib.py (hunk 19 above)
Comment (JelleZijlstra) [reply Lib/test/test_annotationlib.py:257]: I'll leave metaclasses to a separate PR (#119180).

**Comment by JelleZijlstra** [comment]
Thanks @carljm for the review! I pushed fixes for the issues you identified.
